README.md

```diff
@@ -1,3 +1,97 @@
+   gets its name from the statistical method of estimating parameters from
+   certain statistical distributions (e.g. Gaussian statistics correspond to the
+   standard *L2*-estimator, etc).
+3. A high-level optimization routine that performs iteratively reweighted least-squares,
+   solving the effectively non-linear optimization of the cost/loss function by
+   a series of quadratic minimization steps.
+
+### Solvers
+
+The package has three in-built solvers - *QR* factorization of the ordinary
+least-squares problem, the normal-form of the least-squares problem, and conjugate
+gradient on the former.
+
+We are trying to find *x* such that *|r|* is minimized where the residual *r = Ax - b* (i.e. find the *x* so
+that *Ax* is as close as possible to *b* - if *A* is invertible, the solution is
+*inv(A).b* and *r = 0*). The equivalent "normal" form is *A'Ax - A'b*, where *A'A* is now
+a square, Hermitian matrix, and depending on the dimensions of *A* and the
+conditioning of the systems, this may be easier or harder to solve.
+
+A *weighted* least-squares problem introduces a weight vector *w* which is
+multiplies *r* element-wise to scale some residual elements to be larger or smaller,
+increasing or decreasing their importance in the optimization procedure (which
+is what lets us reduce the impact of outliers on the solution).
+
+The solver is called using the `solve()` function
+
+```julia
+x = solve(A, b, weights, method, x0)
+#    A: M x N matrix (dense, sparse, etc)
+#    b: M-vector
+#    weights: scaling factors to be applied to each element of *r*
+#    method: a symbol, one of :qr, :normal or :cg
+#    x0: initial starting vector, for use with :cg method.
+```
+
+### M-Estimators
+
+But how does one obtain the optimal weight vector for reducing the impact of
+outliers? An M-estimator (typically) dampens the impact of the largest elements
+of *r*, making them impact the solution less than in the typical *L2*-estimator
+where the cost function is `sum(r.^2)`. See [this](http://research.microsoft.com/en-us/um/people/zhang/INRIA/Publis/Tutorial-Estim/node24.html) excellent resource for an
+overview.
+
+Included *M*-estimators are represented by Julia instances of subtypes of `MEstimator`
+
+1. `L2Estimator()`, *cost* = `sum(r.^2)`
+2. `L1Estimator()`, *cost* = `sum(abs(r))`
+3. `L1L2Estimator()`, behaves *L2* for *r << 1*, and *L1* for *r >> 1*
+4. `FairEstimator(c)`, behaves *L2* for *r << c*, and *L1* for *r >> c*
+5. `HuberEstimator(c)`, behaves *L2* for *r << c*, and *L1* for *r >> c*
+6. `CauchyEstimator(c)`, behaves *L2* for *r << c*, and logarithmic for *r >> c* (non-convex)
+7. `GemanEstimator()`,  behaves *L2* for *r << 1*, and Lorentzian for *r >> 1* (non-convex)
+8. `WelschEstimator(c)`, behaves *L2* for *r << c*, and Gaussian for *r >> c* (non-convex)
+9. `TukeyEstimator(c)`, behaves *L2* for *r < c*, and complete cutoff for *r > c* (non-convex)
```

Comment:
Nice summary with the asymptotics listed.  Github markdown rendering seems broken for some reason.